Fixes #4971 - Simplify Connection.upgradeFrom()/upgradeTo().

sbordet opened the pull request:
Now the upgrade-from connection produces a "floating" buffer (not belonging to a pool), so that it can release the original buffer.

The upgrade-to connection is free to copy or store this "floating" buffer.

Updated javadocs and all implementations.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Fixes #4971 - Simplify Connection.upgradeFrom()/upgradeTo().

Now the upgrade-from connection produces a "floating" buffer (not belonging to a pool), so that it can release the original buffer.

The upgrade-to connection is free to copy or store this "floating" buffer.

Updated javadocs and all implementations.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Fixes #4971 - Simplify Connection.upgradeFrom()/upgradeTo().

Flip those buffers!

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Fixes #4971 - Simplify Connection.upgradeFrom()/upgradeTo().

Fixed ConnectHandler by calling fillInterested() if the upgrade buffer is null.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

lorban commented on jetty-io/src/main/java/org/eclipse/jetty/io/Connection.java:
```diff
@@ -96,31 +96,49 @@
          */
-        void onUpgradeTo(ByteBuffer prefilled);
+        void onUpgradeTo(ByteBuffer buffer);
     }
```
The contract changes you're proposing forces a `ByteBuffer` allocation.

Couldn't that be avoided by specifying that the buffer returned by `onUpgradeFrom` and passed to `onUpgradeTo` must be fully consumed and no reference to it should be kept?

I would phrase that as:

For `onUpgradeTo`: Implementations MUST consume all bytes of the buffer and MUST NOT keep a reference to the buffer.
For `onUpgradeFrom`: The buffer WILL only be used for the duration of the `Endpoint.upgrade` call and MAY be disposed of afterwards.

sbordet replied:
@lorban and I discussed this and the problem with the approach he proposes is that `onUpgradeFrom()` returns a buffer, but it won't know when it's free to release it back to the pool.

lorban commented on jetty-io/src/main/java/org/eclipse/jetty/io/EndPoint.java:
```diff
@@ -266,13 +266,15 @@
+     * <p>If the current connection is an instance of {@link Connection.UpgradeFrom} then
+     * a buffer of unconsumed bytes is requested; if there are unconsumed bytes, and if
+     * the new connection is an instance of {@link Connection.UpgradeTo}, the unconsumed
+     * buffer is passed to the new connection.</p>
```
Shouldn't we also specify what happens when the current connection is not an instance of `UpgradeFrom` and/or the new connection is not an instance of `UpgradeTo`?

Also, we could specify what happens when `onUpgradeTo` does not consume all the bytes of the buffer, like throwing an exception, or discarding the bytes with a warning.

lorban reviewed (commented):
I find the contract clarified by upgradeFrom()/upgradeTo() not ideal, so I proposed some further changes.

But I suspect you're going to say we cannot do that for backward compatibility reasons, which is fair. In that case, maybe my proposal should be done in 10.x?

sbordet commented:
After discussion with @lorban, the proposed changes may break implementers that would have returned the `onUpgradeTo()` buffer to the pool.

This change should be rescheduled for `jetty-10.0.x`.

sbordet commented:
@gregw what do you think of the comment above?
Should we do this change in 9.4.x and risk to break those very few that have implemented a custom `Connection` _and_ do upgrades, but tidy up the code for the future, or should we just make this change for 10.0.x (and still breaking)?

gregw commented:
@sbordet I think it is fine to break this API contract.  It is highly unlikely that anybody has implemented a connection and supports upgrade. Even if they did, a robust bufferpool would either reject or just use an arbitrary buffer passed to it.  Just in case we should make sure our buffer pools do handle a random buffer released to them

sbordet committed:
Fixes #4971 - Simplify Connection.upgradeFrom()/upgradeTo().

Strengthened ByteBufferPool behavior when releasing non-pooled
ByteBuffers: the buffer is now discarded.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

lorban reviewed (approved):
LGTM

sbordet committed:
Fixes #4971 - Simplify Connection.upgradeFrom()/upgradeTo().

Javadocs.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Fixes #4971 - Simplify Connection.upgradeFrom()/upgradeTo().

Fixed javadocs.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>
